src/CrvInfoHelper.cc

```diff
@@ -0,0 +1,263 @@
+      const CRSScintillatorBarIndex &barIndex = (*iter)->index();
+      for(int SiPM=0; SiPM<4; SiPM++)
+      {
+        if(!(*iter)->getBarDetail().hasCMB(SiPM%2)) continue;
+        sipm_map[barIndex.asInt()*4 + SiPM] = iSiPM;
+        iSiPM++;
+      }
+    }
+
+    // Loop through all reco pulses
+    for(size_t recoPulseIndex=0; recoPulseIndex<crvRecoPulses->size(); recoPulseIndex++)
+    {
+      const art::Ptr<CrvRecoPulse> crvRecoPulse(crvRecoPulses, recoPulseIndex);
+      //get information about the counter
+      const CRSScintillatorBarIndex &barIndex = crvRecoPulse->GetScintillatorBarIndex();
+      int sectorNumber  = -1;
+      int moduleNumber  = -1;
+      int layerNumber   = -1;
+      int counterNumber = -1;
+      CrvHelper::GetCrvCounterInfo(CRS, barIndex, sectorNumber, moduleNumber, layerNumber, counterNumber);
+
+      //Reco pulses information
+      int SiPM = crvRecoPulse->GetSiPMNumber();
+      int SiPMId = sipm_map.find(barIndex.asInt()*4 + SiPM)->second;
+      CLHEP::Hep3Vector HitPos = CrvHelper::GetCrvCounterPos(CRS, barIndex);
+      recoInfo.emplace_back(HitPos, barIndex.asInt(), sectorNumber, SiPMId,
+          crvRecoPulse->GetPEs(), crvRecoPulse->GetPEsPulseHeight(), crvRecoPulse->GetPulseHeight(),
+          crvRecoPulse->GetPulseBeta(), crvRecoPulse->GetPulseFitChi2(), crvRecoPulse->GetPulseTime());
+
+      //MCtruth pulses information
+      double visibleEnergyDeposited  = 0;
+      double earliestHitTime         = 0;
+      double avgHitTime         = 0;
+      CLHEP::Hep3Vector earliestHitPos;
+      CLHEP::Hep3Vector avgHitPos;
+      art::Ptr<SimParticle> mostLikelySimParticle;
+      //for this reco pulse
+      CrvMCHelper::GetInfoFromCrvRecoPulse(crvRecoPulse, crvDigiMCs, visibleEnergyDeposited,earliestHitTime, earliestHitPos, avgHitTime, avgHitPos,
+          mostLikelySimParticle);
+
+      bool hasMCInfo = (mostLikelySimParticle.isNonnull()?true:false); //MC
+      if(hasMCInfo)
+      {
+        const art::Ptr<SimParticle> &primaryParticle = FindPrimaryParticle(mostLikelySimParticle);
+        const art::Ptr<SimParticle> &parentParticle = FindParentParticle(mostLikelySimParticle);
+        const art::Ptr<SimParticle> &gparentParticle = FindGParentParticle(mostLikelySimParticle);
+        MCInfo.emplace_back(true, mostLikelySimParticle->pdgId(),
+            primaryParticle->pdgId(),
+            primaryParticle->startMomentum().e() - primaryParticle->startMomentum().m(),
+            tdet->toDetector(primaryParticle->startPosition()),
+            parentParticle->pdgId(),
+            parentParticle->startMomentum().e()  - parentParticle->startMomentum().m(),
+            tdet->toDetector(parentParticle->startPosition()),
+            gparentParticle->pdgId(),
+            gparentParticle->startMomentum().e() - gparentParticle->startMomentum().m(),
+            tdet->toDetector(gparentParticle->startPosition()),
+            tdet->toDetector(earliestHitPos),
+            earliestHitTime,
```

Comment:
should be replaced by avgHitPos and avgHitTime

Reply:
will do